### PR TITLE
Implementation of redux-toolkit

### PR DESCRIPTION
**Pull request by Untouchamp**

### CODE REVIEW

File: src/app/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { useDispatch, useSelector } from 'react-redux';
```
Comment (Gudashev) [inline src/app/hooks.ts:1]: rename app folder to store

File: src/components/ToDoList/TodoWrapper.tsx [hunk 2]
```diff
@@ -1,21 +1,29 @@
...
 
-function TodoWrapper({ todoTasks }: TodoWrapperType): JSX.Element {
-    const [tasks, addTask, deleteTodo, updateTask]: CustomTypeHooksType =
-        useCustomTaskHooks(todoTasks);
+function TodoWrapper(): JSX.Element {
+    const [addTask, deleteTask, updateTask]: CustomTypeHooksType =
+        useCustomTaskHooks();
+
+    const dispatch = useAppDispatch();
+    const { todos }: Array<TaskType> = useAppSelector((state) => state.todos);
```
Comment (Gudashev) [inline src/components/ToDoList/TodoWrapper.tsx:14]: create selector in store, then useSelector(tasks)

File: src/features/todoList/TodosSlicer.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import { TaskType, UpdateTaskParams } from '../../components/ToDoList/types';
+import uuid from 'uuidv4';
+
+export const getTodosAsync = createAsyncThunk<Array<TaskType>>(
+    'todos/getTodosAsync',
+    async () => {
+        const response = await fetch('http://localhost:7000/todos');
+        if (response.ok) {
+            const todos = response.json();
+            console.log(await todos);
```
Comment (Gudashev) [inline src/features/todoList/TodosSlicer.ts:12]: remove console log

File: src/features/todoList/TodosSlicer.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import type { PayloadAction } from '@reduxjs/toolkit';
+import { TaskType, UpdateTaskParams } from '../../components/ToDoList/types';
+import uuid from 'uuidv4';
+
+export const getTodosAsync = createAsyncThunk<Array<TaskType>>(
+    'todos/getTodosAsync',
+    async () => {
+        const response = await fetch('http://localhost:7000/todos');
+        if (response.ok) {
+            const todos = response.json();
```
Comment (Gudashev) [inline src/features/todoList/TodosSlicer.ts:11]: const todos  = await  response.json();

File: src/features/todoList/TodosSlicer.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+const initialState: TodosState = {
+    todos: [],
+};
+
+export const todosSlice = createSlice({
+    name: 'todos',
+    initialState,
+    reducers: {
+        addTodo: (state: TodosState, action: PayloadAction<string>) => {
+            const newTodoList = {
```
Comment (Gudashev) [inline src/features/todoList/TodosSlicer.ts:82]: move this to service level

File: src/features/todoList/TodosSlicer.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                id: uuid(),
+                description: action.payload,
+                isCompleted: false,
+                isEditing: false,
+            };
+            state.todos.push(newTodoList);
+        },
+        deleteTodo: (state: TodosState, action: PayloadAction<number>) => {
+            state.todos = state.todos.filter(
```
Comment (Gudashev) [inline src/features/todoList/TodosSlicer.ts:91]: https://redux-toolkit.js.org/api/createEntityAdapter, use EntityAdapter API

File: src/features/todoList/TodosSlicer.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    }
+);
+
+// Define a type for the slice state
+interface TodosState {
+    todos: Array<TaskType>;
+}
+
+// Define the initial state using that type
+const initialState: TodosState = {
```
Comment (Gudashev) [inline src/features/todoList/TodosSlicer.ts:73]: [EntityAdapter](https://redux-toolkit.js.org/usage/usage-guide#normalizing-with-createentityadapter)

File: src/hooks/CustomTaskHooks.ts [hunk 8]
```diff
@@ -27,24 +24,10 @@ function useCustomTaskHooks(todoTasks: Array<TaskType>) {
...
-            const updatedTasks = [...tasks];
-            updatedTasks[taskIndex] = updatedTask;
-
-            setTasks(updatedTasks);
-        }
+        dispatch(updateTodoAsync({ id, taskText, isCompleted, isEditing }));
     };
 
-    return [tasks, addTask, deleteTodo, updateTask];
+    return [addTask, deleteTask, updateTask];
```
Comment (Gudashev) [inline src/hooks/CustomTaskHooks.ts:30]: get tasks from  the store in the hook

File: src/components/ToDoList/TodoWrapper.tsx (hunk 2 above)
Comment (Untouchamp) [reply src/components/ToDoList/TodoWrapper.tsx]: In [docs](https://redux-toolkit.js.org/tutorials/typescript) they recommend to create selector in a separate folder called hooks
Or I didn't get your instructions, what the actual problem?

File: src/components/ToDoList/TodoWrapper.tsx [hunk 9]
```diff
@@ -1,21 +1,28 @@
...
 
-function TodoWrapper({ todoTasks }: TodoWrapperType): JSX.Element {
-    const [tasks, addTask, deleteTodo, updateTask]: CustomTypeHooksType =
-        useCustomTaskHooks(todoTasks);
+function TodoWrapper(): JSX.Element {
+    const [todos, addTask, deleteTask, updateTask]: CustomTypeHooksType =
+        useCustomTaskHooks();
+
+    const dispatch = useAppDispatch();
+    useEffect(() => {
```
Comment (Gudashev) [inline src/components/ToDoList/TodoWrapper.tsx:14]: move to hook

File: src/hooks/CustomTaskHooks.ts [hunk 10]
```diff
@@ -1,24 +1,29 @@
...
-            ...tasks,
-            {
-                id: uuid(),
-                description: task,
-                isCompleted: false,
-                isEditing: false,
-            },
-        ]);
+        const todo = {
+            id: uuid(),
```
Comment (Gudashev) [inline src/hooks/CustomTaskHooks.ts:17]: Usually the backend service generates the ID. Transmit without ID, and generate it at the service level

File: src/store/services/todoServices.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import { TaskType, UpdateTaskParams } from '../../components/ToDoList/types';
+
+const todoServices = {
+    getAllTodosService: async () => {
+        const response = await fetch('http://localhost:7000/todos');
```
Comment (Gudashev) [inline src/store/services/todoServices.ts:6]: wrap async code in trycatch

File: src/store/services/todoServices.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        if (response.ok) {
```
Comment (Gudashev) [inline src/store/services/todoServices.ts:7]: remove this, use trycatch

File: src/components/ToDoList/TodoWrapper.tsx (hunk 2 above)
Comment (Gudashev) [reply src/components/ToDoList/TodoWrapper.tsx]: Okay, you can move them to the hooks fodder.

**Review by Gudashev — APPROVED** [review]

**Pull request merged**